File: tests/drills.py
"""T1 / T2 / T6 -- the three tests the architecture rests on (PROTOTYPE.md §3).

Each has a pre-registered kill criterion. Exit code is non-zero if any fails.
"""
from __future__ import annotations

import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from brain import okf                                     # noqa: E402
from brain.cli import iter_concept_files, load_all        # noqa: E402

PY = sys.executable
RESULTS: list[tuple[str, bool, str]] = []


def record(name: str, ok: bool, detail: str = "") -> None:
    RESULTS.append((name, ok, detail))
    print(f"{'PASS' if ok else 'FAIL'}  {name}" + (f"  -- {detail}" if detail else ""))


def sh(*a, cwd=None):
    return subprocess.run(a, cwd=cwd, capture_output=True, text=True)


def brain(bundle: Path, *a):
    return sh(PY, "-m", "brain", "--bundle", str(bundle), *a,
              cwd=str(Path(__file__).resolve().parents[1]))


# --------------------------------------------------------------------- T1

def t1_rebuild(bundle: Path) -> None:
    """KILL: any generated index.md differing by one byte after a full rebuild."""
    brain(bundle, "index", "-q")
    before = {p.relative_to(bundle).as_posix(): p.read_bytes()
              for p in bundle.rglob("index.md")}

    # nuke everything derived
    sh("rm", "-rf", str(bundle / ".brain"))
    for p in list(bundle.rglob("index.md")):
        p.unlink()

    r = brain(bundle, "index", "-q")
    if r.returncode != 0:
        record("T1 rebuild drill", False, f"index failed: {r.stderr.strip()[:200]}")
        return

    after = {p.relative_to(bundle).as_posix(): p.read_bytes()
             for p in bundle.rglob("index.md")}
    if set(before) != set(after):
        missing = set(before) ^ set(after)
        record("T1 rebuild drill", False, f"index.md set changed: {sorted(missing)[:5]}")
        return
    diff = [k for k in before if before[k] != after[k]]
    record("T1 rebuild drill", not diff,
           "byte-identical across full rebuild" if not diff
           else f"{len(diff)} file(s) differ: {diff[:3]}")


def t1_idempotent(bundle: Path) -> None:
    """A second index run must change nothing."""
    brain(bundle, "index", "-q")
    a = {p: p.read_bytes() for p in bundle.rglob("index.md")}
    brain(bundle, "index", "-q")
    b = {p: p.read_bytes() for p in bundle.rglob("index.md")}
    diff = [str(k) for k in a if a[k] != b[k]]
    record("T1b idempotence", not diff,
           "second run is a no-op" if not diff else f"{len(diff)} changed")


# --------------------------------------------------------------------- T2

def t2_raw_roundtrip(bundle: Path, tmp: Path) -> None:
    """KILL: a single non-identical byte through ingest -> export."""
    brain(bundle, "index", "-q")
    r = brain(bundle, "export", str(tmp), "--force")
    if r.returncode != 0:
        record("T2a raw round-trip", False, r.stderr.strip()[:200])
        return
    bad = []
    for full, rel in iter_concept_files(bundle):
        out = tmp / rel
        if not out.exists():
            bad.append(f"{rel} (missing)")
        elif out.read_bytes() != full.read_bytes():
            bad.append(rel)
    record("T2a raw round-trip", not bad,
           "byte-identical incl. CRLF, unicode paths, anchors"
           if not bad else f"{len(bad)} differ: {bad[:3]}")


def t2_semantic_roundtrip(bundle: Path) -> None:
    """Would a schema-backed store lose anything? Parse -> canonical reserialize
    -> reparse, and compare the PARSED VALUES. Formatting may change; data must not.
    This is the diagnostic for the Operator's silent-truncation failure mode."""
    lost = []
    for c in load_all(bundle):
        if not c.fm:
            continue
        try:
            re_text = okf.format_concept(c)
            c2 = okf.parse(c.path, re_text)
        except Exception as e:
            lost.append(f"{c.path}: reserialize raised {type(e).__name__}")
            continue
        missing = set(c.fm) - set(c2.fm)
        if missing:
            lost.append(f"{c.path}: lost keys {sorted(missing)}")
            continue
        for k, v in c.fm.items():
            if c2.fm.get(k) != v:
                lost.append(f"{c.path}: key `{k}` changed value")
                break
        if c2.body.strip() != c.body.strip():
            lost.append(f"{c.path}: body changed")
    record("T2b semantic round-trip", not lost,
           "no key or value lost through reserialization"
           if not lost else f"{len(lost)} loss(es): {lost[:3]}")


def t2_unknown_keys(bundle: Path) -> None:
    """Spec: consumers MUST NOT reject unknown types/keys, and MUST tolerate
    broken links. Verify we actually indexed them rather than dropping them."""
    import json as _json
    import sqlite3
    db = bundle / ".brain" / "index.db"
    con = sqlite3.connect(db); con.row_factory = sqlite3.Row
    row = con.execute("SELECT fm_json,type FROM concepts WHERE path=?",
                      ("adversarial/unknown-type.md",)).fetchone()
    if not row:
        record("T2c unknown keys/types", False, "concept not indexed at all")
        return
    fm = _json.loads(row["fm_json"])
    problems = []
    if row["type"] != "ThreatModel":
        problems.append("unknown type not preserved")
    for k in ("retention", "classification"):
        if k not in fm:
            problems.append(f"unknown key `{k}` dropped")
    broken = con.execute("SELECT COUNT(*) c FROM edges WHERE resolved=0").fetchone()["c"]
    if broken == 0:
        problems.append("broken links not recorded")
    record("T2c unknown keys/types", not problems,
           f"unknown type + 2 unknown keys preserved, {broken} broken links recorded"
           if not problems else "; ".join(problems))


def t2_no_type_coercion(bundle: Path) -> None:
    """REGRESSION: PyYAML's implicit timestamp resolver silently rewrites
    `2026-01-01T00:00:00Z` to `2026-01-01 00:00:00+00:00`, which is not ISO 8601.
    Nothing in a parsed concept may be a date/datetime object."""
    import datetime as _dt
    bad = []

    def walk(v, path):
        if isinstance(v, (_dt.date, _dt.datetime)):
            bad.append(path)
        elif isinstance(v, dict):
            for k, vv in v.items():
                walk(vv, f"{path}.{k}")
        elif isinstance(v, list):
            for i, vv in enumerate(v):
                walk(vv, f"{path}[{i}]")

    for c in load_all(bundle):
        walk(c.fm, c.path)
    record("T2d no timestamp coercion", not bad,
           "all temporal values remain ISO 8601 strings"
           if not bad else f"{len(bad)} coerced to datetime: {bad[:3]}")


def t2_fmt_is_safe(bundle: Path) -> None:
    """`brain fmt` is the only command that rewrites a file. Every rewrite it
    would perform must preserve parsed values exactly, and it must REFUSE files
    whose authored structure it cannot reproduce (anchors, block scalars)."""
    unsafe, refused = [], 0
    for c in load_all(bundle):
        reason = okf.formattable(c)
        if reason:
            refused += 1
            continue
        out = okf.format_concept(c)
        c2 = okf.parse(c.path, out)
        if c2.fm != c.fm or c2.body.strip() != c.body.strip():
            unsafe.append(c.path)
    record("T2e fmt safety", not unsafe,
           f"every rewrite value-preserving; {refused} file(s) correctly refused"
           if not unsafe else f"{len(unsafe)} unsafe rewrite(s): {unsafe[:3]}")


# --------------------------------------------------------------------- T6

def t6_service_death(bundle: Path, tmp: Path) -> None:
    """Destroy everything derived and the index db, then rebuild from the bare
    bundle. KILL: concept count differs by >0."""
    brain(bundle, "index", "-q")
    import sqlite3
    db = bundle / ".brain" / "index.db"
    con = sqlite3.connect(db)
    before = con.execute("SELECT COUNT(*) FROM concepts").fetchone()[0]
    con.close()

    sh("rm", "-rf", str(bundle / ".brain"))
    for p in list(bundle.rglob("index.md")):
        p.unlink()
    r = brain(bundle, "index", "-q")
    if r.returncode != 0:
        record("T6 service-death drill", False, r.stderr.strip()[:200])
        return
    con = sqlite3.connect(db)
    after = con.execute("SELECT COUNT(*) FROM concepts").fetchone()[0]
    con.close()
    record("T6 service-death drill", before == after,
           f"{after} concepts recovered from files alone (was {before})")


# -------------------------------------------------------------------- search

def t_search_smoke(bundle: Path) -> None:
    r = brain(bundle, "search", "revocation", "--limit", "3")
    ok = r.returncode == 0 and "revocation" in r.stdout.lower()
    record("search smoke", ok, (r.stdout.strip().splitlines() or ["no output"])[0])

    r = brain(bundle, "search", "index", "--type", "Decision", "--limit", "5")
    ok2 = r.returncode == 0 and "decisions/" in r.stdout
    record("search structured filter", ok2, "type=Decision filter applied")


if __name__ == "__main__":
    root = Path(__file__).resolve().parents[1]
    # Optional bundle argument. The self-authored fixture is the default, but FINDINGS notes it is
    # "adversarial but self-authored, which is its weakness" -- real OKF written by people who did not
    # know our assumptions is the stronger test. Without this the drills could only ever run against
    # our own fixture, which is why T2-against-upstream had never been run.
    bundle = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else root / "_fixture"
    tmp = root / ("_export" if len(sys.argv) < 2 else f"_export-{bundle.name}")
    if not bundle.exists():
        print(f"no such bundle: {bundle}")
        print("usage: python tests/drills.py [BUNDLE_DIR]   (default: _fixture)")
        print("build the fixture first: python tests/make_fixture.py _fixture")
        sys.exit(2)

    print("=" * 66)
    print("STAGE 1 DRILLS")
    print("=" * 66)
    t1_rebuild(bundle)
    t1_idempotent(bundle)
    t2_raw_roundtrip(bundle, tmp)
    t2_semantic_roundtrip(bundle)
    t2_unknown_keys(bundle)
    t2_no_type_coercion(bundle)
    t2_fmt_is_safe(bundle)
    t6_service_death(bundle, tmp)
    t_search_smoke(bundle)

    failed = [n for n, ok, _ in RESULTS if not ok]
    print("=" * 66)
    print(f"{len(RESULTS) - len(failed)}/{len(RESULTS)} passed")
    if failed:
        print("FAILED: " + ", ".join(failed))
    sys.exit(1 if failed else 0)

File: okf-conformance/tests/drills.rs
//! The universal drills: every one runs against every bundle.
//!
//! Each has a pre-registered kill criterion from PROTOTYPE.md §3. These are the assertions the
//! architecture rests on — if one fails, a claim in ARCHITECTURE.md's evidence table is false,
//! not merely a test being red.
//!
//! Failures accumulate across bundles and are reported together. A gate that stops at the first
//! bundle tells you one thing is broken; this one tells you whether it is broken *everywhere*,
//! which is usually the question you actually have.

use okf_conformance::*;

/// Defects FINDINGS.md has already adjudicated and nobody has yet decided how to fix.
///
/// `(drill, bundle, substring the failure must contain, why)`.
///
/// A recorded defect is not an excuse — it is a claim with two edges, and both are checked.
/// If the failure changes shape, that is a regression. If it *stops happening*, that is also
/// reported, because it means someone fixed a defect the record still calls open and FINDINGS
/// is now lying. Silence in either direction is how a known-issues list rots into an ignore-list.
const KNOWN_DEFECTS: &[(&str, &str, &str, &str)] = &[(
    "T1 rebuild",
    "acme_retail",
    "attesters/index.md",
    "E4b -- `attesters/` holds no concepts, so the index.md upstream ships for it cannot be \
     regenerated. A1 (\"everything above the bundle is derived\") is falsified as stated: either \
     index rebuilds concept-free directories, or A1 narrows to directories that contain concepts. \
     Undecided -- see FINDINGS.md E4b.",
)];

fn known_defect(drill: &str, bundle: &str) -> Option<(&'static str, &'static str)> {
    KNOWN_DEFECTS
        .iter()
        .find(|(d, b, _, _)| *d == drill && *b == bundle)
        .map(|(_, _, needle, why)| (*needle, *why))
}

/// Run `f` over every bundle, collecting failures rather than panicking on the first.
///
/// Reporting every bundle matters: a gate that stops at the first failure tells you something is
/// broken; this one tells you whether it is broken *everywhere*, which is usually the question
/// you actually have.
fn for_each_bundle(drill: &str, f: impl Fn(&Bundle) -> Result<String, String>) {
    let mut failures = Vec::new();
    let mut notes = Vec::new();
    let bundles = all_bundles();
    assert!(!bundles.is_empty(), "no bundles found -- the suite would be vacuous");

    for src in &bundles {
        let b = Bundle::checkout(src);
        let expected = known_defect(drill, &b.name);
        match (f(&b), expected) {
            (Ok(note), None) => notes.push(format!("  PASS   {:<24} {note}", b.name)),
            (Ok(_), Some((_, why))) => {
                let msg = format!(
                    "recorded defect no longer reproduces -- FINDINGS.md must be updated to close it.\n         {why}"
                );
                failures.push(format!("  FIXED  {:<24} {msg}", b.name));
                notes.push(format!("  FIXED  {:<24} {msg}", b.name));
            }
            (Err(why), Some((needle, note))) if why.contains(needle) => {
                notes.push(format!("  XFAIL  {:<24} {why}\n         {note}", b.name));
            }
            (Err(why), Some((needle, _))) => {
                let msg = format!("known defect changed shape (expected it to mention `{needle}`): {why}");
                failures.push(format!("  FAIL   {:<24} {msg}", b.name));
                notes.push(format!("  FAIL   {:<24} {msg}", b.name));
            }
            (Err(why), None) => {
                failures.push(format!("  FAIL   {:<24} {why}", b.name));
                notes.push(format!("  FAIL   {:<24} {why}", b.name));
            }
        }
    }
    println!("{drill}\n{}", notes.join("\n"));
    assert!(failures.is_empty(), "\n{drill}\n{}\n", failures.join("\n"));
}

// ── T1 ──────────────────────────────────────────────────────────────────────

/// KILL: any generated `index.md` differing by one byte after a full rebuild.
///
/// This is the drill the whole "derived and disposable" claim rests on. If it fails, the
/// derived plane is not derived — it holds state nothing else can reconstruct.
#[test]
fn t1_rebuild_is_byte_identical() {
    for_each_bundle("T1 rebuild", |b| {
        b.ok(&["index", "-q"]);
        let before = b.index_files();
        // Comparing two empty sets is a pass that asserts nothing. An implementation that
        // generates no index.md at all would otherwise sail through this drill.
        if before.is_empty() {
            return Err("no index.md generated -- nothing to compare, drill would pass vacuously".into());
        }

        b.destroy_derived();
        b.ok(&["index", "-q"]);
        let after = b.index_files();

        if before.keys().ne(after.keys()) {
            let lost: Vec<_> = before.keys().filter(|k| !after.contains_key(*k)).collect();
            let gained: Vec<_> = after.keys().filter(|k| !before.contains_key(*k)).collect();
            return Err(format!("index.md set changed -- lost {lost:?}, gained {gained:?}"));
        }
        let diff: Vec<&String> = before.iter().filter(|(k, v)| after.get(*k) != Some(v)).map(|(k, _)| k).collect();
        if diff.is_empty() {
            Ok(format!("{} index.md byte-identical across full rebuild", before.len()))
        } else {
            Err(format!("{} file(s) differ: {:?}", diff.len(), &diff[..diff.len().min(3)]))
        }
    });
}

/// A second index run must change nothing. Non-idempotence means the index is a function of
/// its own prior state, not of the files — which would quietly break T1 on someone else's machine.
#[test]
fn t1b_index_is_idempotent() {
    for_each_bundle("T1b idempotence", |b| {
        // Destroy first, so the files being compared were definitely produced by THIS run.
        // Without it the drill reads the `index.md` files the upstream bundles vendor, and an
        // implementation that does nothing at all looks perfectly idempotent.
        b.destroy_derived();
        b.ok(&["index", "-q"]);
        let a = b.index_files();
        if a.is_empty() {
            return Err("no index.md generated -- idempotence over nothing is not idempotence".into());
        }
        b.ok(&["index", "-q"]);
        let c = b.index_files();
        if a == c {
            Ok(format!("second run is a no-op ({} index.md unchanged)", a.len()))
        } else {
            let diff: Vec<&String> = a.iter().filter(|(k, v)| c.get(*k) != Some(v)).map(|(k, _)| k).collect();
            Err(format!("{} file(s) changed on re-run: {:?}", diff.len(), diff))
        }
    });
}

// ── T2 ──────────────────────────────────────────────────────────────────────

/// KILL: a single non-identical byte through ingest → export.
///
/// The claim being defended is that `export` writes *raw bytes*, so no serializer sits in the
/// write path where it could drop an unknown key. That makes the failure mode structurally
/// impossible rather than merely untested — but only if this drill actually holds.
#[test]
fn t2a_round_trip_is_byte_exact() {
    for_each_bundle("T2a raw round-trip", |b| {
        b.ok(&["index", "-q"]);
        let out = b.root.parent().unwrap().join("export");
        let out_s = out.to_string_lossy().into_owned();
        b.ok(&["export", &out_s, "--force"]);

        let mut bad = Vec::new();
        for (rel, bytes) in b.concept_files() {
            let exported = out.join(&rel);
            match std::fs::read(&exported) {
                Err(_) => bad.push(format!("{rel} (missing)")),
                Ok(got) if got != bytes => bad.push(rel),
                Ok(_) => {}
            }
        }
        if bad.is_empty() {
            Ok("byte-identical incl. CRLF, unicode paths, anchors".into())
        } else {
            Err(format!("{} differ: {:?}", bad.len(), &bad[..bad.len().min(3)]))
        }
    });
}

/// T2b — no key or value lost through canonical reserialization.
///
/// `fmt` is the only command that rewrites a concept, so it is the only place a value could be
/// silently lost. Parse → canonicalize → reparse, and compare the PARSED VALUES: formatting is
/// allowed to change, data is not. This is the diagnostic for the silent-truncation failure mode.
#[test]
fn t2b_reserialization_loses_nothing() {
    for_each_bundle("T2b semantic round-trip", |b| {
        b.ok(&["index", "-q"]);
        let paths: Vec<String> = b.concept_files().into_keys().collect();
        let before: Vec<serde_json::Value> = paths.iter().map(|p| b.show(p)).collect();

        b.ok(&["fmt"]);

        let mut lost = Vec::new();
        for (rel, was) in paths.iter().zip(&before) {
            let now = b.show(rel);
            if was["frontmatter"] != now["frontmatter"] {
                lost.push(format!("{rel}: frontmatter changed"));
            } else if was["trust"] != now["trust"] {
                // The tier is derived from frontmatter, so this cannot drift on its own --
                // if it does, the derivation is reading something the round-trip destroyed.
                lost.push(format!("{rel}: trust tier flipped {} -> {}", was["trust"], now["trust"]));
            }
        }
        if lost.is_empty() {
            Ok(format!("{} concepts survive canonical rewrite unchanged", paths.len()))
        } else {
            Err(format!("{} loss(es): {:?}", lost.len(), &lost[..lost.len().min(3)]))
        }
    });
}

/// T2d — no temporal value coerced off ISO 8601.
///
/// REGRESSION GUARD (FINDINGS E3a). PyYAML's implicit timestamp resolver silently rewrites
/// `2026-01-01T00:00:00Z` into `2026-01-01 00:00:00+00:00`, which is *not* ISO 8601 — a
/// space where the `T` belongs. That defect is why the frontmatter parser is a port at all.
///
/// Checked two ways. The first is universal and catches the exact PyYAML shape. The second is
/// exact but only safe on concepts the implementation considers reproducible: inside anchors,
/// merge keys and block scalars a token in the raw text legitimately need not appear verbatim
/// in the parsed view.
#[test]
fn t2d_temporal_values_stay_iso8601() {
    for_each_bundle("T2d no timestamp coercion", |b| {
        b.ok(&["index", "-q"]);
        let mut bad = Vec::new();
        let mut checked = 0usize;

        for (rel, raw) in b.concept_files() {
            let doc = b.show(&rel);
            let fm = &doc["frontmatter"];
            let mut leaves = Vec::new();
            walk_json(fm, "", &mut leaves);

            // (a) universal: no value may be a timestamp rewritten off ISO 8601. Matched on the
            // whole value, not on any date appearing inside it -- prose mentions dates.
            for (path, v) in &leaves {
                if let Some(s) = v.as_str() {
                    if is_coerced_timestamp(s) {
                        bad.push(format!("{rel}{path}: `{s}` is not ISO 8601 (space separator)"));
                    }
                }
            }

            // (b) exact, where the file is safely reproducible.
            if doc["formattable_refusal"].is_null() {
                checked += 1;
                let fm_str = serde_json::to_string(fm).unwrap_or_default();
                for tok in iso8601_tokens(&frontmatter_text(&raw)) {
                    if !fm_str.contains(&tok) {
                        bad.push(format!("{rel}: `{tok}` did not survive parsing verbatim"));
                    }
                }
            }
        }
        if bad.is_empty() {
            Ok(format!("all temporal values remain ISO 8601 strings ({checked} verbatim-checked)"))
        } else {
            Err(format!("{} coerced: {:?}", bad.len(), &bad[..bad.len().min(3)]))
        }
    });
}

/// T2e — the formatter is safe.
///
/// `fmt` is the only command that rewrites a file, so it is the only one that can destroy
/// authored content. Two obligations: it must REFUSE files whose structure it cannot reproduce,
/// and having refused them it must leave them byte-untouched.
///
/// This drill exists because building the formatter revealed the danger (FINDINGS E3b): the
/// naive canonicalizer resolved merge keys, invented an `&id001` anchor on an unrelated
/// timestamp, and flattened a shell script under `script: |` into a quoted string. Without the
/// drill that damage ships silently.
#[test]
fn t2e_formatter_refuses_what_it_cannot_reproduce() {
    for_each_bundle("T2e fmt safety", |b| {
        b.ok(&["index", "-q"]);

        // Which files does the implementation itself say it must refuse?
        let refused: Vec<String> = b
            .concept_files()
            .into_keys()
            .filter(|rel| !b.show(rel)["formattable_refusal"].is_null())
            .collect();
        let before = b.concept_files();

        b.ok(&["fmt"]);
        let after = b.concept_files();

        let mut violations = Vec::new();
        for rel in &refused {
            if before.get(rel) != after.get(rel) {
                violations.push(format!("{rel}: refused by fmt but rewritten anyway"));
            }
        }
        // `fmt --check` must now be clean: a formatter that cannot reach a fixed point would
        // make every CI run dirty and the check meaningless.
        let check = b.run(&["fmt", "--check"]);
        if check.code != 0 {
            let would: Vec<&str> = check.stdout.lines().filter(|l| l.starts_with("would reformat")).collect();
            violations.push(format!("fmt is not a fixed point: {:?}", &would[..would.len().min(3)]));
        }

        if violations.is_empty() {
            Ok(format!("{} file(s) correctly refused and left untouched", refused.len()))
        } else {
            Err(violations.join("; "))
        }
    });
}

// ── T6 ──────────────────────────────────────────────────────────────────────

/// KILL: concept count differs by more than 0 after destroying everything derived.
///
/// The asymmetry the whole design leans on: losing the derived plane costs time, not knowledge.
/// This is the drill that makes that a measurement rather than a slogan.
#[test]
fn t6_index_recovers_from_files_alone() {
    for_each_bundle("T6 service-death", |b| {
        b.ok(&["index", "-q"]);
        let before = b.concept_count();

        b.destroy_derived();
        b.ok(&["index", "-q"]);
        let after = b.concept_count();

        if before == after {
            Ok(format!("{after} concepts recovered from files alone"))
        } else {
            Err(format!("{after} recovered, expected {before}"))
        }
    });
}

// ── Conformance floor ───────────────────────────────────────────────────────

/// The spec requires consumers not to reject unknown types, unknown keys, or broken links.
/// Rejection is not the only way to fail that: silently dropping a concept fails it too, which
/// is exactly the defect E4a found. Every concept file present must be indexed.
#[test]
fn every_concept_file_is_indexed() {
    for_each_bundle("conformance floor", |b| {
        b.ok(&["index", "-q"]);
        let on_disk = b.concept_files().len();
        let indexed = b.concept_count();
        if on_disk == indexed {
            Ok(format!("{indexed} concepts, none dropped"))
        } else {
            Err(format!("{on_disk} concept files on disk but {indexed} indexed -- {} dropped", on_disk as i64 - indexed as i64))
        }
    });
}
